### ISSUE TITLE
Custom list of default casters and serializers

### ISSUE THREAD

**Issue by osteel** [EventSaucePHP/ObjectHydrator#48]
Hi,

I've just migrated my events to using the object mapper and while it works great, I'd like my domain events to be completely free of infrastructure details, which I consider the `PropertyCaster` and `PropertySerializer` attributes to be a part of.

Looking at [`DefaultCasterRepository`](https://github.com/EventSaucePHP/ObjectHydrator/blob/main/src/DefaultCasterRepository.php) and [`DefaultSerializerRepository`](https://github.com/EventSaucePHP/ObjectHydrator/blob/main/src/DefaultSerializerRepository.php) I can see that a bunch of default casters and serializers are registered.

E.g. in `DefaultCasterRepository`:

```php
public static function builtIn(): static
{
    $repository = new static();
    $repository->registerDefaultCaster(DateTimeImmutable::class, CastToDateTimeImmutable::class);
    $repository->registerDefaultCaster(DateTimeInterface::class, CastToDateTimeImmutable::class);
    $repository->registerDefaultCaster(UuidInterface::class, CastToUuid::class);

     return $repository;
}
```

The object mapper will automatically cast/serialize objects of these types using the corresponding casters/serializers, without the need for attributes.

It'd be great to be able to extend those lists of default casters/serializers, which would allow us to do so from a config file, for instance (similar principle as EventSauce's [`ExplicitlyMappedClassNameInflector`](https://github.com/EventSaucePHP/EventSauce/blob/main/src/ExplicitlyMappedClassNameInflector.php)).

I've tried to implement this but can't see any obvious extension points. The lowest non-final class down the chain is `ObjectMapperUsingReflection`, and it requires a `DefinitionProvider` – marked as final – itself requiring a `DefaultCasterRepository`, also final.

Could those not be final anymore? And/or some interfaces be introduced for both `DefinitionProvider` and `DefaultCasterRepository` (and potentially `DefaultSerializerRepository`) to be used in `ObjectMapperUsingReflection` (and the new `DefinitionProvider` interface) in place of the final concrete implementations?

It would then allow us to inject a different `DefinitionProvider` in `ObjectMapperUsingReflection`, with custom repositories for casters and serializers.

Basically, these `final` keywords and lack of interfaces are the only things preventing us from using custom objects at the moment – the classes are otherwise already designed so these object can be injected easily.

Cheers,

Yannick

**Comment by osteel** [comment]
I've got a working implementation [here](https://github.com/osteel/dime/pull/42) for reference.

For it to work I duplicated the `DefinitionProvider` class, which I've customised slightly.

I've also copied `ObjectMapperUsingReflection` into `ObjectMapperUsingReflectionAndClassMap` and tweaked it a little.

The latter has the former as a dependency.

I also have two `CasterRepository` and `SerializerRepository` classes, used by my copy of `DefinitionProvider`.

It works, but instead of having quasi identical copies of `ObjectMapperUsingReflection` and `DefinitionProvider` (because I'm still interested in most of their logic and behaviour), I'd rather be able to just extend them and inject my own `CasterRepository` and `SerializerRepository`.

**Comment by frankdejonge** [comment]
@osteel have you considered constructing your own default caster/serializer repository and configuring using the register methods? You can pass it before passing it into the definition provider? A config file based approach would be more for a framework binding library than the core package.

**Comment by osteel** [comment]
You're right – pushed the new version [here](https://github.com/osteel/dime/pull/42/files). It's now using a [factory](https://github.com/osteel/dime/pull/42/files#diff-da1f967ca0f3ea1a7cb0cdc63145ffe0271aec72a689b2e86548f41ef8958ec0) to build the payload serialiser. The factory receives the property-to-hydrator class map array.

As for the config-based approach, I wasn't advocating for adding it to the core package – in my case, the config file management is provided by Laravel already, and really what I wanted was to inject the property-to-hydrator class map somewhere. The corresponding array could have come from anywhere – config file, database, etc. – without it being the core package's concern.

Anyway, that solves my issue – thank you for the pointer!
